iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1879,70 +1908,58 @@ static void processIncomingMessageNotification(PMQTTTRANSPORT_HANDLE_DATA transp
     else if (extractMqttProperties(transportData, IoTHubMessage, topicName, type) != 0)
     {
         LogError("failure extracting mqtt properties.");
+        IoTHubMessage_Destroy(IoTHubMessage);
     }
     else
     {
-        messageData = (MESSAGE_CALLBACK_INFO*)malloc(sizeof(MESSAGE_CALLBACK_INFO));
-        if (messageData == NULL)
+        MESSAGE_DISPOSITION_CONTEXT* dispositionContext = createMessageDispositionContext(msgHandle);
+
+        if (dispositionContext == NULL)
         {
-            LogError("malloc failed");
+            LogError("failed creating message disposition context");
+            IoTHubMessage_Destroy(IoTHubMessage);
+        }
+        else if (IoTHubMessage_SetDispositionContext(IoTHubMessage, dispositionContext, destroyMessageDispositionContext) != IOTHUB_MESSAGE_OK)
+        {
+            LogError("Failed setting disposition context in IOTHUB_MESSAGE_HANDLE");
+            destroyMessageDispositionContext(dispositionContext);
+            IoTHubMessage_Destroy(IoTHubMessage);
         }
         else
         {
-            messageData->messageHandle = IoTHubMessage;
-            messageData->transportContext = NULL;
-
             if (type == IOTHUB_TYPE_EVENT_QUEUE)
             {
                 // Codes_SRS_IOTHUB_MQTT_TRANSPORT_31_065: [ If type is IOTHUB_TYPE_TELEMETRY and sent to an input queue, then on success `mqttNotificationCallback` shall call `IoTHubClient_LL_MessageCallback`. ]
-                if (!transportData->transport_callbacks.msg_input_cb(messageData, transportData->transport_ctx))
+                if (!transportData->transport_callbacks.msg_input_cb(IoTHubMessage, transportData->transport_ctx))
                 {
-                    // Cleanup of allocated memory happens at end of function as pointers are non-NULL.
                     LogError("IoTHubClientCore_LL_MessageCallbackFromInput returned false");
-                }
-                else
-                {
+                    // This will destroy the dispostion context;
+                    IoTHubMessage_Destroy(IoTHubMessage);
                     IoTHubMessage = NULL;
```

Comment:
Please remove this `IoTHubMessage = NULL` and the one in block a bit below.  This was needed the way code was previously structured where a non-NULL would get freed in closing block, but not now.

Reply:
Hmmm. Indeed. Thanks.